File: android-native/KyouNoOgatore/app/src/main/java/jp/ogatore/kyouno/card/BragCardRenderer.kt
package jp.ogatore.kyouno.card

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.DashPathEffect
import android.graphics.LinearGradient
import android.graphics.Paint
import android.graphics.RectF
import android.graphics.Shader

// ネイティブ移植 Step 7b(マスタープラン§2-1「index.html drawBragCard」行・§6 Step 7b): じまんカード
// 描画(index.html:2805-2919 drawBragCard)の1:1移植。CardRendererとは独立の描画器として作業量を
// 見積もる、という§2-1備考どおりファイル自体は分けているが、実際の背景・飾り・白カード・タイトルピル・
// 日付バッジの舞台演出はdrawCard()と全く同じ数値(index.html:2814のコメント「記録カードと同じ舞台」)
// なので、その部分の図形/テキスト/色ヘルパーはCardRenderer(internal公開済み)を呼んで再利用する。
//
// Step4のCardRendererと同じ簡略化方針を踏襲: キャラクター立ち絵・M PLUS 1p/BananaNumフォント
// (Typeface.DEFAULT_BOLDで代替)は未実装。動画サムネイルはネットワーク取得(https://i.ytimg.com/...)を
// 要するため、常にWeb版の「オフラインでサムネイルが出せないとき」の代替パス(動画タイトルを2行まで
// 折り返し表示)を採用する——ネットワーク依存を増やさない方針(既存のsdVideoHTML等と同じ判断)。
//
// 現在時刻・乱数を直接読まない設計(§1-1第3項・§2-4末尾の禁止事項。厳守): フッターメッセージの選定は
// 日付文字列の31進ハッシュ(index.html:2902と同一のVoicesLogic.dailyHash相当をこのファイル内で直接計算)
// のみで決まり、システム乱数源・現在時刻APIには一切触れない。
object BragCardRenderer {
    // index.html:2897-2901 pool の1:1移植。
    private val footerPool = listOf(
        "続けてるじぶん、どんどんじまんしてね✨",
        "この1本と続けた日々が、もうじまんです。",
        "続けてるあなたが、いちばんすてきです。",
    )

    // index.html:2818-2820 deco の1:1移植(drawCard側の固定配置と全く同じ13件。§2-1備考どおり
    // 「記録カードと同じ舞台」のため値を重複させず流用してもよいが、drawBragCard自体が常にこの
    // 固定配置のみを使う=CARD_THEMES_V2_FROM分岐が存在しない仕様なので、ここに独立して持つ)。
    private data class Deco(val shape: String, val x: Float, val y: Float, val sz: Float)
    private val fixedDeco = listOf(
        Deco("h", 95f, 150f, 34f), Deco("s", 885f, 120f, 26f), Deco("k", 120f, 860f, 30f), Deco("h", 905f, 850f, 30f),
        Deco("c", 60f, 480f, 11f), Deco("k", 940f, 430f, 24f), Deco("s", 180f, 70f, 18f), Deco("c", 500f, 60f, 9f),
        Deco("h", 820f, 300f, 22f), Deco("c", 935f, 650f, 10f), Deco("s", 70f, 690f, 16f), Deco("c", 250f, 935f, 9f), Deco("k", 760f, 945f, 22f),
    )

    // index.html:2808 の1:1移植(小数入力を弾かないtype=numberの実測バグ修正込み)。
    fun clampDays(raw: Int): Int = raw.coerceIn(1, 9999)

    fun render(ds: String, days: Int, theme: ResolvedTheme, favoriteTitle: String?): Bitmap {
        val bitmap = Bitmap.createBitmap(1000, 1000, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        draw(canvas, ds, clampDays(days), theme, favoriteTitle)
        return bitmap
    }

    private fun draw(canvas: Canvas, ds: String, days: Int, theme: ResolvedTheme, favoriteTitle: String?) {
        val R = CardRenderer // ヘルパー呼び出しの見た目短縮用

        // 背景グラデ+固定飾り(index.html:2815-2827。記録カードと同じ舞台)
        val bgPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            shader = LinearGradient(0f, 0f, 1000f, 1000f, R.color(theme.bg[0]), R.color(theme.bg[1]), Shader.TileMode.CLAMP)
        }
        canvas.drawRect(0f, 0f, 1000f, 1000f, bgPaint)
        fixedDeco.forEachIndexed { i, d ->
            val col = R.color(theme.deco[i % theme.deco.size])
            when (d.shape) {
                "h" -> R.drawHeartShape(canvas, d.x - d.sz / 2, d.y - d.sz / 2, d.sz, col)
                "s" -> R.drawStarShape(canvas, d.x, d.y, d.sz * 0.6f, col)
                "k" -> R.drawSparkleShape(canvas, d.x, d.y, d.sz * 0.7f, col)
                else -> canvas.drawCircle(d.x, d.y, d.sz, Paint(Paint.ANTI_ALIAS_FLAG).apply { color = col })
            }
        }

        // 白カード+破線ふち(index.html:2828-2830。drawCardと同じ数値)
        canvas.drawRoundRect(RectF(85f, 175f, 915f, 825f), 52f, 52f, Paint(Paint.ANTI_ALIAS_FLAG).apply { color = Color.argb(240, 255, 255, 255) })
        val borderPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            style = Paint.Style.STROKE; strokeWidth = 4f; strokeCap = Paint.Cap.ROUND
            color = R.colorWithAlpha(theme.main, 0.45f)
            pathEffect = DashPathEffect(floatArrayOf(2f, 16f), 0f)
        }
        canvas.drawRoundRect(RectF(110f, 200f, 890f, 800f), 40f, 40f, borderPaint)

        // タイトルピル(index.html:2832-2834。drawCardと同じ数値)
        canvas.drawRoundRect(RectF(300f, 145f, 700f, 209f), 32f, 32f, Paint(Paint.ANTI_ALIAS_FLAG).apply { color = R.color(theme.main) })
        R.drawCenteredText(canvas, "#きょうのオガトレ", 500f, 190f, 34f, Color.WHITE)

        // 日付バッジ(index.html:2836-2841。drawCardと同じ数値)
        val parts = ds.split("-").mapNotNull { it.toIntOrNull() }
        val dtxt = if (parts.size == 3) "${parts[0]}/${parts[1]}/${parts[2]}" else ds
        val dw = R.textWidth(dtxt, 26f)
        canvas.drawRoundRect(RectF(868f - dw - 44f, 212f, 868f, 264f), 26f, 26f, Paint(Paint.ANTI_ALIAS_FLAG).apply { color = R.colorWithAlpha(theme.main, 0.85f) })
        R.drawCenteredText(canvas, dtxt, 868f - (dw + 44f) / 2f, 247f, 26f, Color.WHITE)

        // つづけてる日数(index.html:2843-2856。桁数でフォントサイズを変える。BananaNum未同梱のためBOLDで代替)
        val numTxt = days.toString()
        val numSize = if (numTxt.length <= 2) 200f else if (numTxt.length == 3) 170f else 140f
        val numW = R.textWidth(numTxt, numSize)
        val sw = R.textWidth("日つづいてる！", 52f)
        val totalW = numW + 18f + sw
        val startX = 500f - totalW / 2f
        R.drawLeftText(canvas, numTxt, startX, 438f, numSize, R.color(theme.main))
        R.drawLeftText(canvas, "日つづいてる！", startX + numW + 18f, 428f, 52f, R.color("#3A3A35"))

        // 区切り線(index.html:2859-2860)
        val dividerPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            style = Paint.Style.STROKE; strokeWidth = 3f
            color = R.colorWithAlpha(theme.main, 0.3f)
            pathEffect = DashPathEffect(floatArrayOf(4f, 12f), 0f)
        }
        canvas.drawLine(170f, 488f, 830f, 488f, dividerPaint)

        // 「すきな1本」タグピル(index.html:2862-2875)
        run {
            val label = "すきな1本"
            val lw = R.textWidth(label, 28f)
            val pw = lw + 48f
            val yc = 525f
            canvas.drawRoundRect(RectF(500f - pw / 2f, yc - 30f, 500f - pw / 2f + pw, yc + 30f), 30f, 30f, Paint(Paint.ANTI_ALIAS_FLAG).apply { color = R.color(theme.main) })
            R.drawLeftText(canvas, label, 500f - pw / 2f + 24f, yc + 10f, 28f, Color.WHITE)
        }

        // サムネイル代替=動画タイトルの折り返し表示(index.html:2883-2889。ネットワーク非依存のため常にこの経路)
        val favT = favoriteTitle ?: "まだえらんでません（これから見つけます！）"
        val lines = wrapLines(favT, 540f, 2)
        lines.forEachIndexed { i, ln -> R.drawCenteredText(canvas, ln, 500f, 645f + i * 52f, 34f, R.color("#3A3A35")) }

        // フッター=キャラの吹き出し(index.html:2896-2914。キャラ本体は未実装のため吹き出しのみ)
        var fh = 0u
        for (c in ds) fh = fh * 31u + c.code.toUInt()
        val fmsg = footerPool[(fh % footerPool.size.toUInt()).toInt()]
        var ffs = 27f
        while (R.textWidth(fmsg, ffs) > 560f && ffs > 21f) ffs -= 1f
        val bw = R.textWidth(fmsg, ffs) + 56f
        val bx1 = maxOf(70f, 690f - bw)
        canvas.drawRoundRect(RectF(bx1, 900f, bx1 + bw, 974f), 37f, 37f, Paint(Paint.ANTI_ALIAS_FLAG).apply { color = Color.argb(242, 255, 255, 255) })
        R.drawLeftText(canvas, fmsg, bx1 + 28f, 900f + 37f + (ffs * 0.36f), ffs, R.color(theme.main))
    }

    // index.html:2792-2804(drawBragCard直前の後勝ち定義)の1:1移植。1文字ずつ幅を測って折り返し、
    // maxLines到達時は末尾を"…"に置き換える。
    private fun wrapLines(text: String, maxW: Float, maxLines: Int): List<String> {
        val lines = mutableListOf<String>()
        var cur = ""
        for (ch in text) {
            val test = cur + ch
            if (CardRenderer.textWidth(test, 34f) > maxW && cur.isNotEmpty()) {
                lines.add(cur)
                cur = ch.toString()
                if (lines.size == maxLines) break
            } else {
                cur = test
            }
        }
        if (lines.size < maxLines && cur.isNotEmpty()) {
            lines.add(cur)
        } else if (lines.size == maxLines && cur.isNotEmpty()) {
            lines[maxLines - 1] = lines[maxLines - 1].dropLast(1) + "…"
        }
        return lines
    }
}

File: ios-native/KyouNoOgatore/CardCore/Sources/CardCore/BragCardRenderer.swift
import CoreGraphics
import CoreText
import Foundation
import ImageIO

// ネイティブ移植 Step 7b(マスタープラン§2-1「index.html drawBragCard」行・§6 Step 7b)→Step4/7b
// パリティ突合タスク(TASK-C2-2026-07-26-native-migration-card-visual-assets.md): じまんカード描画
// (index.html:2805-2919 drawBragCard)の1:1移植(Android版BragCardRenderer.ktと同一ロジック)。
// CardRendererとは独立の描画器として作業量を見積もる、という§2-1備考どおりファイル自体は分けているが、
// 実際の背景・飾り・白カード・タイトルピル・日付バッジの舞台演出はdrawCard()と全く同じ数値
// (index.html:2814のコメント「記録カードと同じ舞台」)なので、その部分の図形/テキスト/色ヘルパーは
// CardRenderer(internal公開済み・同一パッケージ)を呼んで再利用する。キャラクター立ち絵・
// M PLUS 1p/BananaNumフォントもCardRenderer側の実装(CHARA_FILES/CardFonts)をそのまま再利用する
// (§2-1備考「じまん・声…」行の「同じアセット・フォント」)。
//
// TASK-C2-2026-07-27-brag-card-thumbnail.md: index.html:2876-2889 drawBragCard()のサムネイル分岐の
// 1:1移植。取得(3秒タイムアウト・失敗時null)自体はネットワークI/Oのため、CardCore(決定的ロジックの
// 1:1移植先)の外側=アプリ層(BragView.swift)で行い、ここには結果のCGImageだけを受け取る。
// thumbnailがnilのとき(未選択・オフライン・タイムアウト)は従来どおり動画タイトルの折り返し表示
// (index.html:2883-2889)へフォールバックする(このフォールバック自体は変更しない)。
public enum BragCardRenderer {
    private static let footerPool = [
        "続けてるじぶん、どんどんじまんしてね✨",
        "この1本と続けた日々が、もうじまんです。",
        "続けてるあなたが、いちばんすてきです。",
    ]

    private static let fixedDeco: [(String, CGFloat, CGFloat, CGFloat)] = [
        ("h", 95, 150, 34), ("s", 885, 120, 26), ("k", 120, 860, 30), ("h", 905, 850, 30),
        ("c", 60, 480, 11), ("k", 940, 430, 24), ("s", 180, 70, 18), ("c", 500, 60, 9),
        ("h", 820, 300, 22), ("c", 935, 650, 10), ("s", 70, 690, 16), ("c", 250, 935, 9), ("k", 760, 945, 22),
    ]

    // index.html:2808 の1:1移植(小数入力を弾かないtype=numberの実測バグ修正込み)。
    public static func clampDays(_ raw: Int) -> Int { min(9999, max(1, raw)) }

    public static func render(ds: String, days: Int, theme: ResolvedTheme, favoriteTitle: String?, thumbnail: CGImage? = nil) -> Data {
        let width = 1000, height = 1000
        let colorSpace = CGColorSpaceCreateDeviceRGB()
        guard let ctx = CGContext(
            data: nil, width: width, height: height, bitsPerComponent: 8, bytesPerRow: 0,
            space: colorSpace, bitmapInfo: CGImageAlphaInfo.noneSkipLast.rawValue
        ) else { return Data() }

        ctx.translateBy(x: 0, y: CGFloat(height))
        ctx.scaleBy(x: 1, y: -1)

        draw(in: ctx, ds: ds, days: clampDays(days), theme: theme, favoriteTitle: favoriteTitle, thumbnail: thumbnail)

        guard let image = ctx.makeImage() else { return Data() }
        return CardRenderer.pngData(from: image)
    }

    private static func draw(in ctx: CGContext, ds: String, days: Int, theme: ResolvedTheme, favoriteTitle: String?, thumbnail: CGImage?) {
        // 背景グラデ+固定飾り(index.html:2815-2827。記録カードと同じ舞台)
        let colorSpace = CGColorSpaceCreateDeviceRGB()
        let colors = [CardRenderer.color(theme.bg[0]), CardRenderer.color(theme.bg[1])] as CFArray
        if let gradient = CGGradient(colorsSpace: colorSpace, colors: colors, locations: [0, 1]) {
            ctx.drawLinearGradient(gradient, start: .zero, end: CGPoint(x: 1000, y: 1000), options: [])
        }
        for (i, d) in fixedDeco.enumerated() {
            let col = CardRenderer.color(theme.deco[i % theme.deco.count])
            switch d.0 {
            case "h": CardRenderer.drawHeartShape(ctx, x: d.1 - d.3 / 2, y: d.2 - d.3 / 2, s: d.3, color: col)
            case "s": CardRenderer.drawStarShape(ctx, x: d.1, y: d.2, r: d.3 * 0.6, color: col)
            case "k": CardRenderer.drawSparkleShape(ctx, x: d.1, y: d.2, r: d.3 * 0.7, color: col)
            default:
                ctx.setFillColor(col)
                ctx.fillEllipse(in: CGRect(x: d.1 - d.3, y: d.2 - d.3, width: d.3 * 2, height: d.3 * 2))
            }
        }

        // 白カード+破線ふち(index.html:2828-2830。drawCardと同じ数値)
        ctx.setFillColor(gray: 1, alpha: 0.94)
        CardRenderer.roundRectPath(ctx, x: 85, y: 175, w: 830, h: 650, r: 52)
        ctx.fillPath()
        ctx.saveGState()
        ctx.setStrokeColor(CardRenderer.color(theme.main, alpha: 0.45))
        ctx.setLineWidth(4)
        ctx.setLineDash(phase: 0, lengths: [2, 16])
        ctx.setLineCap(.round)
        CardRenderer.roundRectPath(ctx, x: 110, y: 200, w: 780, h: 600, r: 40)
        ctx.strokePath()
        ctx.restoreGState()

        // タイトルピル(index.html:2832-2834。drawCardと同じ数値)
        ctx.setFillColor(CardRenderer.color(theme.main))
        CardRenderer.roundRectPath(ctx, x: 300, y: 145, w: 400, h: 64, r: 32)
        ctx.fillPath()
        CardRenderer.drawCenteredText("#きょうのオガトレ", in: ctx, centerX: 500, baselineY: 190, fontSize: 34, weight: .w900, color: CardRenderer.color("#FFFFFF"))

        // 日付バッジ(index.html:2836-2841。drawCardと同じ数値)
        let parts = ds.split(separator: "-").map { Int($0) ?? 0 }
        let dtxt = parts.count == 3 ? "\(parts[0])/\(parts[1])/\(parts[2])" : ds
        let dw = CardRenderer.textWidth(dtxt, fontSize: 26, weight: .w800)
        ctx.saveGState()
        ctx.setAlpha(0.85)
        ctx.setFillColor(CardRenderer.color(theme.main))
        CardRenderer.roundRectPath(ctx, x: 868 - dw - 44, y: 212, w: dw + 44, h: 52, r: 26)
        ctx.fillPath()
        ctx.restoreGState()
        CardRenderer.drawCenteredText(dtxt, in: ctx, centerX: 868 - (dw + 44) / 2, baselineY: 247, fontSize: 26, weight: .w800, color: CardRenderer.color("#FFFFFF"))

        // つづけてる日数(index.html:2843-2856。桁数でフォントサイズを変える。数字はBananaNum)
        let numTxt = String(days)
        let numSize: CGFloat = numTxt.count <= 2 ? 200 : (numTxt.count == 3 ? 170 : 140)
        let numW = CardRenderer.textWidth(numTxt, fontSize: numSize, weight: .banana)
        let sw = CardRenderer.textWidth("日つづいてる！", fontSize: 52, weight: .w900)
        let totalW = numW + 18 + sw
        let startX = 500 - totalW / 2
        CardRenderer.drawLeftText(numTxt, in: ctx, x: startX, baselineY: 438, fontSize: numSize, weight: .banana, color: CardRenderer.color(theme.main))
        CardRenderer.drawLeftText("日つづいてる！", in: ctx, x: startX + numW + 18, baselineY: 428, fontSize: 52, weight: .w900, color: CardRenderer.color("#3A3A35"))

        // 区切り線(index.html:2859-2860)
        ctx.saveGState()
        ctx.setStrokeColor(CardRenderer.color(theme.main, alpha: 0.3))
        ctx.setLineWidth(3)
        ctx.setLineDash(phase: 0, lengths: [4, 12])
        ctx.move(to: CGPoint(x: 170, y: 488))
        ctx.addLine(to: CGPoint(x: 830, y: 488))
        ctx.strokePath()
        ctx.restoreGState()

        // 「すきな1本」タグピル(index.html:2862-2875)
        do {
            let label = "すきな1本"
            let lw = CardRenderer.textWidth(label, fontSize: 28, weight: .banana)
            let pw = lw + 48
            let yc: CGFloat = 525
            ctx.setFillColor(CardRenderer.color(theme.main))
            CardRenderer.roundRectPath(ctx, x: 500 - pw / 2, y: yc - 30, w: pw, h: 60, r: 30)
            ctx.fillPath()
            CardRenderer.drawLeftText(label, in: ctx, x: 500 - pw / 2 + 24, baselineY: yc + 10, fontSize: 28, weight: .banana, color: CardRenderer.color("#FFFFFF"))
        }

        // サムネイル(index.html:2876-2883)。取得できなかったとき(オフライン・タイムアウト・
        // 動画未選択)は従来どおり動画タイトルの折り返し表示(index.html:2883-2889)にフォールバックする。
        if let thumbnail {
            let tw: CGFloat = 416, thh: CGFloat = 234, tx: CGFloat = 500 - 416 / 2, ty: CGFloat = 562
            ctx.saveGState()
            CardRenderer.roundRectPath(ctx, x: tx, y: ty, w: tw, h: thh, r: 18)
            ctx.clip()
            CardRenderer.drawImageJS(thumbnail, x: tx, y: ty, w: tw, h: thh, in: ctx)
            ctx.restoreGState()
            ctx.saveGState()
            ctx.setStrokeColor(CardRenderer.color(theme.main, alpha: 0.5))
            ctx.setLineWidth(3)
            CardRenderer.roundRectPath(ctx, x: tx, y: ty, w: tw, h: thh, r: 18)
            ctx.strokePath()
            ctx.restoreGState()
        } else {
            let favT = favoriteTitle ?? "まだえらんでません（これから見つけます！）"
            let lines = wrapLines(favT, maxW: 540, maxLines: 2)
            for (i, ln) in lines.enumerated() {
                CardRenderer.drawCenteredText(ln, in: ctx, centerX: 500, baselineY: 645 + CGFloat(i) * 52, fontSize: 34, weight: .w800, color: CardRenderer.color("#3A3A35"))
            }
        }

        // キャラ(index.html:2891-2894。日替わりローテ・CardRenderer.CHARA_FILESを共用。§2-1備考どおり
        // 「同じアセット・フォント」を使う。dateIdx駆動の選定理由はCardRenderer.swift冒頭コメント参照)。
        let dateIdx = CardLottery.dateIdx(ds)
        let charaFiles = CardDataLoader.shared.CHARA_FILES
        if !charaFiles.isEmpty {
            let pick = charaFiles[((dateIdx % charaFiles.count) + charaFiles.count) % charaFiles.count]
            if let charaImage = CardRenderer.loadBundleImage(CardRenderer.charaDrawableName(pick.file)) {
                let w: CGFloat = 255
                let h = w * CGFloat(charaImage.height) / CGFloat(charaImage.width)
                CardRenderer.drawImageJS(charaImage, x: 965 - w, y: 985 - h, w: w, h: h, in: ctx)
            }
        }

        // フッター=キャラの吹き出し(index.html:2896-2914)
        var fh: UInt32 = 0
        for u in ds.utf16 { fh = fh &* 31 &+ UInt32(u) }
        let fmsg = footerPool[Int(fh % UInt32(footerPool.count))]
        var ffs: CGFloat = 27
        while CardRenderer.textWidth(fmsg, fontSize: ffs, weight: .w800) > 560 && ffs > 21 { ffs -= 1 }
        let bw = CardRenderer.textWidth(fmsg, fontSize: ffs, weight: .w800) + 56
        let bx1 = max(70, 690 - bw)
        ctx.setFillColor(gray: 1, alpha: 0.95)
        CardRenderer.roundRectPath(ctx, x: bx1, y: 900, w: bw, h: 74, r: 37)
        ctx.fillPath()
        CardRenderer.drawLeftText(fmsg, in: ctx, x: bx1 + 28, baselineY: 900 + 37 + (ffs * 0.36), fontSize: ffs, weight: .w800, color: CardRenderer.color(theme.main))
    }

    // index.html:2792-2804(drawBragCard直前の後勝ち定義)の1:1移植。1文字ずつ幅を測って折り返し、
    // maxLines到達時は末尾を"…"に置き換える。
    private static func wrapLines(_ text: String, maxW: CGFloat, maxLines: Int) -> [String] {
        var lines: [String] = []
        var cur = ""
        for ch in text {
            let test = cur + String(ch)
            if CardRenderer.textWidth(test, fontSize: 34, weight: .w800) > maxW && !cur.isEmpty {
                lines.append(cur)
                cur = String(ch)
                if lines.count == maxLines { break }
            } else {
                cur = test
            }
        }
        if lines.count < maxLines && !cur.isEmpty {
            lines.append(cur)
        } else if lines.count == maxLines && !cur.isEmpty {
            lines[maxLines - 1] = String(lines[maxLines - 1].dropLast()) + "…"
        }
        return lines
    }
}
